File: src/benchmark/framework/DecisionTracer.js
/**
 * Decision Trace Graph — tracks causal chains across AI pipeline phases.
 *
 * Instruments Perceiver → Planner → Executor → Evaluator interfaces
 * and performs backward attribution when negative events occur.
 */

const PHASES = ['perceiver', 'planner', 'executor', 'evaluator', 'director'];

/**
 * A single trace entry recording phase input/output at a point in time.
 */
// { tick, phase, input, output, metadata }

export class DecisionTracer {
  constructor() {
    this._traces = [];       // All trace entries
    this._faultCounts = {};  // { phase: count }
    for (const p of PHASES) this._faultCounts[p] = 0;
    this._negativeEvents = []; // Detected negative events
  }

  /**
   * Record a phase's input/output at a given tick.
   * @param {number} tick
   * @param {string} phase - One of PHASES
   * @param {object} input - What this phase received
   * @param {object} output - What this phase produced
   * @param {object} [metadata] - Extra context
   */
  record(tick, phase, input, output, metadata = {}) {
    this._traces.push({ tick, phase, input, output, metadata, timestamp: Date.now() });
  }

  /**
   * Record a negative event (population decline, resource depletion, etc).
   * @param {number} tick
   * @param {string} type - 'population_decline' | 'resource_depletion' | 'objective_failure'
   * @param {object} details
   */
  recordNegativeEvent(tick, type, details) {
    this._negativeEvents.push({ tick, type, details });
  }

  /**
   * Perform backward attribution for a negative event.
   * Walks back through the trace to find which phase is responsible.
   *
   * Attribution logic:
   * 1. Find the most recent traces before the event tick
   * 2. Check executor: did it deviate from planner instructions?
   * 3. Check planner: did it have the crisis signal in its input but ignore it?
   * 4. Check perceiver: did the raw state contain the signal but perceiver missed it?
   *
   * @param {{ tick, type, details }} event
   * @returns {{ attributedPhase: string, confidence: number, reasoning: string }}
   */
  attributeEvent(event) {
    const priorTraces = this._traces.filter(t => t.tick <= event.tick && t.tick >= event.tick - 150);
    // Group by phase, take most recent
    const latest = {};
    for (const t of priorTraces) {
      if (!latest[t.phase] || t.tick > latest[t.phase].tick) latest[t.phase] = t;
    }

    // Check executor first: did it try to execute the plan?
    const executorTrace = latest['executor'];
    const plannerTrace = latest['planner'];
    const perceiverTrace = latest['perceiver'];

    if (executorTrace && plannerTrace) {
      const planSteps = plannerTrace.output?.steps?.length ?? 0;
      const executed = executorTrace.output?.executedSteps ?? 0;
      if (planSteps > 0 && executed === 0) {
        this._faultCounts['executor']++;
        return { attributedPhase: 'executor', confidence: 0.8, reasoning: 'Plan existed but executor completed 0 steps' };
      }
    }

    // Check planner: did it receive crisis signal but not address it?
    if (plannerTrace && perceiverTrace) {
      const hasCrisisSignal = this._detectCrisisSignal(perceiverTrace.output, event.type);
      const planAddressesCrisis = this._planAddressesCrisis(plannerTrace.output, event.type);
      if (hasCrisisSignal && !planAddressesCrisis) {
        this._faultCounts['planner']++;
        return { attributedPhase: 'planner', confidence: 0.7, reasoning: 'Perceiver detected crisis but planner did not address it' };
      }
    }

    // Check perceiver: did it miss the signal entirely?
    if (perceiverTrace) {
      const hasCrisisSignal = this._detectCrisisSignal(perceiverTrace.output, event.type);
      if (!hasCrisisSignal) {
        this._faultCounts['perceiver']++;
        return { attributedPhase: 'perceiver', confidence: 0.6, reasoning: 'Crisis signal not present in perceiver output' };
      }
    }

    // Default: attribute to director for not orchestrating response
    this._faultCounts['director']++;
    return { attributedPhase: 'director', confidence: 0.4, reasoning: 'No specific phase fault identified; director failed to orchestrate' };
  }

  /** Check if perceiver output contains a signal related to the event type */
  _detectCrisisSignal(perceiverOutput, eventType) {
    if (!perceiverOutput) return false;
    switch (eventType) {
      case 'resource_depletion':
        return perceiverOutput.economy?.food?.projectedZeroSec != null
            || perceiverOutput.economy?.food?.stock < 15
            || perceiverOutput.economy?.wood?.stock < 10;
      case 'population_decline':
        return perceiverOutput.workforce?.growthBlockers?.length > 0
            || perceiverOutput.defense?.activeSaboteurs > 0;
      case 'objective_failure':
        return perceiverOutput.objective?.progressPct < 30;
      default: return false;
    }
  }

  /** Check if planner output addresses the crisis type */
  _planAddressesCrisis(plannerOutput, eventType) {
    if (!plannerOutput?.steps) return false;
    const stepTypes = plannerOutput.steps.map(s => s.type?.toLowerCase() ?? '');
    switch (eventType) {
      case 'resource_depletion':
        return stepTypes.some(t => t.includes('farm') || t.includes('lumber') || t.includes('food'));
      case 'population_decline':
        return stepTypes.some(t => t.includes('wall') || t.includes('defense') || t.includes('clinic'));
      case 'objective_failure':
        return stepTypes.some(t => t.includes('warehouse') || t.includes('objective'));
      default: return false;
    }
  }

  /**
   * Run attribution for all recorded negative events.
   * Safe to call multiple times — computes fault counts from scratch each time.
   * @returns {{ faultDistribution: Record<string, number>, attributions: object[] }}
   */
  analyzeAll() {
    // Reset fault counts before re-computing to avoid double-counting on repeated calls
    for (const p of PHASES) this._faultCounts[p] = 0;

    const attributions = this._negativeEvents.map(e => ({
      event: e,
      attribution: this.attributeEvent(e),
    }));
    const total = Object.values(this._faultCounts).reduce((a, b) => a + b, 0) || 1;
    const faultDistribution = {};
    for (const p of PHASES) {
      faultDistribution[p] = Math.round(this._faultCounts[p] / total * 100);
    }
    return { faultDistribution, attributions, totalEvents: this._negativeEvents.length };
  }

  /** Reset all state */
  reset() {
    this._traces = [];
    this._negativeEvents = [];
    for (const p of PHASES) this._faultCounts[p] = 0;
  }
}

File: project_utopia/benchmark/framework/decision_tracer.py
"""Decision Trace Graph (port of ``DecisionTracer.js``).

Tracks Perceiver -> Planner -> Executor -> Evaluator causal chains and
performs backward attribution for negative events. Pure logic; no sim
dependencies.
"""

from __future__ import annotations

import time
from typing import Any

__all__ = ["DecisionTracer", "PHASES"]


PHASES: tuple[str, ...] = ("perceiver", "planner", "executor", "evaluator", "director")


class DecisionTracer:
    """Trace per-phase IO + attribute negative events."""

    __slots__ = ("_traces", "_fault_counts", "_negative_events")

    def __init__(self) -> None:
        self._traces: list[dict[str, Any]] = []
        self._fault_counts: dict[str, int] = {p: 0 for p in PHASES}
        self._negative_events: list[dict[str, Any]] = []

    def record(
        self,
        tick: int,
        phase: str,
        input_payload: Any,
        output_payload: Any,
        metadata: dict[str, Any] | None = None,
    ) -> None:
        self._traces.append(
            {
                "tick": tick,
                "phase": phase,
                "input": input_payload,
                "output": output_payload,
                "metadata": metadata or {},
                "timestamp": int(time.time() * 1000),
            }
        )

    def record_negative_event(self, tick: int, event_type: str, details: Any) -> None:
        self._negative_events.append({"tick": tick, "type": event_type, "details": details})

    def attribute_event(self, event: dict[str, Any]) -> dict[str, Any]:
        prior = [
            t
            for t in self._traces
            if t["tick"] <= event["tick"] and t["tick"] >= event["tick"] - 150
        ]
        latest: dict[str, dict[str, Any]] = {}
        for t in prior:
            if t["phase"] not in latest or t["tick"] > latest[t["phase"]]["tick"]:
                latest[t["phase"]] = t

        executor_trace = latest.get("executor")
        planner_trace = latest.get("planner")
        perceiver_trace = latest.get("perceiver")

        if executor_trace and planner_trace:
            plan_steps = len((planner_trace["output"] or {}).get("steps", []))
            executed = (executor_trace["output"] or {}).get("executedSteps", 0)
            if plan_steps > 0 and executed == 0:
                self._fault_counts["executor"] += 1
                return {
                    "attributedPhase": "executor",
                    "confidence": 0.8,
                    "reasoning": "Plan existed but executor completed 0 steps",
                }

        if planner_trace and perceiver_trace:
            has_crisis = self._detect_crisis_signal(
                perceiver_trace["output"], event["type"]
            )
            plan_addresses = self._plan_addresses_crisis(
                planner_trace["output"], event["type"]
            )
            if has_crisis and not plan_addresses:
                self._fault_counts["planner"] += 1
                return {
                    "attributedPhase": "planner",
                    "confidence": 0.7,
                    "reasoning": "Perceiver detected crisis but planner did not address it",
                }

        if perceiver_trace:
            has_crisis = self._detect_crisis_signal(
                perceiver_trace["output"], event["type"]
            )
            if not has_crisis:
                self._fault_counts["perceiver"] += 1
                return {
                    "attributedPhase": "perceiver",
                    "confidence": 0.6,
                    "reasoning": "Crisis signal not present in perceiver output",
                }

        self._fault_counts["director"] += 1
        return {
            "attributedPhase": "director",
            "confidence": 0.4,
            "reasoning": "No specific phase fault identified; director failed to orchestrate",
        }

    @staticmethod
    def _detect_crisis_signal(perceiver_output: Any, event_type: str) -> bool:
        if not isinstance(perceiver_output, dict):
            return False
        if event_type == "resource_depletion":
            economy = perceiver_output.get("economy") or {}
            food = (economy.get("food") or {}) if isinstance(economy, dict) else {}
            wood = (economy.get("wood") or {}) if isinstance(economy, dict) else {}
            return (
                food.get("projectedZeroSec") is not None
                or (food.get("stock") is not None and float(food.get("stock", 100)) < 15)
                or (wood.get("stock") is not None and float(wood.get("stock", 100)) < 10)
            )
        if event_type == "population_decline":
            workforce = perceiver_output.get("workforce") or {}
            defense = perceiver_output.get("defense") or {}
            blockers = workforce.get("growthBlockers") if isinstance(workforce, dict) else None
            sabs = defense.get("activeSaboteurs", 0) if isinstance(defense, dict) else 0
            return bool(blockers) or (sabs and sabs > 0)
        if event_type == "objective_failure":
            objective = perceiver_output.get("objective") or {}
            if not isinstance(objective, dict):
                return False
            pct = objective.get("progressPct")
            return pct is not None and float(pct) < 30
        return False

    @staticmethod
    def _plan_addresses_crisis(planner_output: Any, event_type: str) -> bool:
        if not isinstance(planner_output, dict) or not planner_output.get("steps"):
            return False
        step_types = [
            (s.get("type", "") or "").lower()
            for s in planner_output["steps"]
            if isinstance(s, dict)
        ]
        if event_type == "resource_depletion":
            return any("farm" in t or "lumber" in t or "food" in t for t in step_types)
        if event_type == "population_decline":
            return any("wall" in t or "defense" in t or "clinic" in t for t in step_types)
        if event_type == "objective_failure":
            return any("warehouse" in t or "objective" in t for t in step_types)
        return False

    def analyze_all(self) -> dict[str, Any]:
        # Reset counts to avoid double-counting on repeated calls.
        for p in PHASES:
            self._fault_counts[p] = 0
        attributions = [
            {"event": e, "attribution": self.attribute_event(e)}
            for e in self._negative_events
        ]
        total = max(1, sum(self._fault_counts.values()))
        fault_distribution = {
            p: round(self._fault_counts[p] / total * 100) for p in PHASES
        }
        return {
            "faultDistribution": fault_distribution,
            "attributions": attributions,
            "totalEvents": len(self._negative_events),
        }

    def reset(self) -> None:
        self._traces.clear()
        self._negative_events.clear()
        for p in PHASES:
            self._fault_counts[p] = 0
